.github/workflows/main.yml

```diff
@@ -0,0 +1,200 @@
+        with:
+          python-version: "3.9"
+          architecture: "x64"
+
+      - name: Install Poetry ${{ matrix.poetry-version }}
+        uses: abatilo/actions-poetry@v2.3.0
+        with:
+          poetry-version: ${{ matrix.poetry-version }}
+
+      - name: Install dependencies
+        run: poetry install --only style
+
+      # check formatting of the code style
+      - name: Check code formatting
+        run: poetry run poe check_format
+
+      # this applies various linting
+      - name: Lint codebase
+        run: poetry run poe lint
+
+      - name: Type check
+        run: poetry run poe type_check
+
+  build:
+    timeout-minutes: 15
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu, macos, windows]
+        python-version: [3.8, 3.9, "3.10"]
+        poetry-version: [1.3.0]
+    name: build ${{ matrix.os }} - py${{ matrix.python-version }}
+    runs-on: ${{ matrix.os }}-latest
+    defaults:
+      run:
+        shell: bash
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+      - name: Setup Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+          architecture: "x64"
+      - name: Install Poetry ${{ matrix.poetry-version }}
+        uses: abatilo/actions-poetry@v2.3.0
+        with:
+          poetry-version: ${{ matrix.poetry-version }}
+      - name: Install from source
+        run: |
+          poetry install
+      - name: Test package install
+        run: poetry run python -c "import pywhy_stats; print(pywhy_stats.__version__)"
+      - name: Remove package install
+        run: python -m pip uninstall -yq pywhy-stats
+      - name: Check poetry lock file
+        run: poetry update --dry-run
+      - name: Build package
+        run: poetry build
+      - name: Upload package distribution files
+        if: ${{ matrix.os == 'ubuntu' && matrix.python-version == '3.10' }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: package
+          path: dist
+
+      #
+      # Test wheel-based Installation
+      #
+      - name: Install wheel
+        run: pip install ./dist/*.whl
+      - name: Test wheel install
+        run: python -c "import pywhy_stats; print(pywhy_stats.__version__)"
+      - name: Remove wheel install
+        run: python -m pip uninstall -yq pywhy-stats
+
+      #
+      # Test sdist-based Installation
+      #
+      - name: Install sdist
+        run: pip install ./dist/*.tar.gz
+      - name: Test sdist install
+        run: python -c "import pywhy_stats; print(pywhy_stats.__version__)"
+      - name: Remove sdist install
+        run: python -m pip uninstall -yq pywhy-stats
+
+  test:
+    timeout-minutes: 30
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu, macos, windows]
+        python-version: [3.8, "3.10"]  # oldest and newest supported versions
```

Comment:
I saw in dodiscover that 3.8 is required, but I didn't fully get why. Shouldn't all these algorithms work with 3.7 as well?

Reply:
Python 3.7 is reaching EOL this year. Prolly best to not support Python versions that aren't supported by PSF: https://pythonspeed.com/articles/stop-using-python-3.7/